### PR TITLE
Experimental test runner (part 2)

### PR DESCRIPTION
**Pull request by jviide**
This pull request continues from pull request #4, adding some major features to showrunner. The set of commits also contains some smaller fixes to Python 2.7 compatibility and `examples/python3-urllib/run.py`.
## On-the-fly Certificate Generation

Add `showrunner.gencert.gencert(cn="...")` for generating a PEM encoded server certificate + private key pair and a PEM ca bundle against which the server certificate verifies. The function basically just wraps the `openssl` command and generates a server certificate for the given Common Name. To save time the private server key and the private CA key are generated once per process.

As an example you can generate `cert.pem`, `cert.key` and `ca.pem` like this:

``` python
from showrunner.gencert import gencert


def writefile(filename, data):
    with open(filename, "wb") as f:
        f.write(data)


cert, key, ca = gencert("localhost")
writefile("cert.pem", cert)
writefile("cert.key", key)
writefile("ca.pem", ca)
```

Now launching a HTTP server on localhost with the `cert.pem` and `cert.key` files allows clients to create a verified connection to it using the `ca.pem` CA file.
## Helpers for Local Testing

Add helpers for writing local test servers: `showrunner.testenv.local(ok_expected, cn, [callback=...])` launches a simple temporary TLS server on `localhost` (with keys generated by `gencert`).  After the test has been completed the temporary TLS server gets shut down.
- `ok_expected` tells whether the connected clients should be able to verify the server certificate.
- `cn` tells the Common Name that should be passed to `gencert` when generating the server certificates.
- The optional `callback` argument can be passed in to modify how a created connection socket is used after a socket connection has been established. The default callback just does the TLS handshake, but is also used by the HTTPS bundle (described later) to imitate a webserver. For an example see [`showrunner.bundles.https.https_callback`](https://github.com/ouspg/trytls/blob/showrunner/showrunner/bundles/https.py#L6-L8).
## Test Bundles

Add the concept of _test bundles_ and add a command line option `-t`/`--test-bundle` to the `trytls` tool for configuring the used test bundle.

A test bundle is a bundle of tests specialized for some protocol/situation ("just a plain TLS handshake", "a HTTPS server", ...). In practice a test bundle is just a named list of tests to be run. As an example see [`showrunner.bundles.handshakes.all_tests`](https://github.com/ouspg/trytls/blob/showrunner/showrunner/bundles/handshake.py#L16) (described later).

The command line parameter `-t` (or `--test-bundle`) is used to tell which test suite should be run. The value should be an absolute Python import path, for example `showrunner.bundles.https.all_tests` for importing the test list `all_tests` from the `showrunner.bundles.https` module. For convenience the path can also be a relative one (i.e. starting with `.`) in which case the bundle is imported under the `showrunner.bundles` package meant for built-in bundles. Therefore `showrunner.bundles.handshake.all_tests` can be referred to as `.handshake.all_tests`.

The default `--test-bundle` value is `.handshake.all_tests`.

The support for absolute import paths is there because that allows people to write their own test bundles that don't have to be a part of the `showrunner` package. This in turn makes it easier to host the bundle code somewhere else than this repository.
### Built-in Bundles

Add some example test bundles of varying quality:
- `.handshake.all_tests`/`showrunner.bundles.handshake.all_tests` for testing plain TLS handshakes (on a local server and on the BadSSL servers). This is the default test bundle for the `trytls` tool.
- `.https.all_tests`/`showrunner.bundles.https.all_tests` that is pretty much like the `.handshake` bundle but also sends a rudimentary HTTP response back to the client. Should help in testing HTTP clients.
- `.imap.all_tests`/`showrunner.bundles.imap.all_tests` that currently just tells to connect to Gmail's IMAP servers.
### Example

Testing the `examples/python3-urllib/run.py` stub against only the local tests in the builtin HTTPS bundle:

``` console
$ trytls -t .https.local_tests python3 examples/python3-urllib/run.py
```

### CODE REVIEW

File: examples/python3-imaplib/run.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+if cafile is None:
+    ssl_context = ssl.create_default_context()
+else:
+    ssl_context = ssl.create_default_context(cafile=cafile)
+
+try:
+    imaplib.IMAP4_SSL(host, port, ssl_context=ssl_context)
+except (ssl.SSLError, ssl.CertificateError):
+    print("FAIL")
```
Comment (evilon) [inline examples/python3-imaplib/run.py:17]: Update these to match the [new return-value spec](https://github.com/ouspg/trytls#return-values)?

**Comment by ikisusi** [comment]
Uh, please consider putting this epic documentation to showrunner/README.md

File: examples/python3-imaplib/run.py (hunk 1 above)
Comment (jviide) [inline examples/python3-imaplib/run.py:17]: It should probably be another PR, as the changes are mainly outside the scope of this pull request.

Comment (jviide) [inline examples/python3-imaplib/run.py:17]: Added #13 to remind about this.
